### DISCUSSION TITLE
Can StrongBranch be used for nonlinear integer optimization?

### DISCUSSION THREAD

**Discussion by cloud-waiting-for-wind** [ds4dm/ecole#275] [Questions]
Can StrongBranch be used for nonlinear integer optimization?
I try to use   ecole.observation.StrongBranchingScores()  on a  nonlinear integer optimization instance,  however,  all entries in the score are 1e-12.  

I'm not sure if  I made a code mistake or ecole.observation.StrongBranchingScores()  doesn't work on nonlinear integer optimization?

**Comment by AntoinePrv** [comment]
From @gasse 

> Hi @cloud-waiting-for-wind ,
>
> In general all of Ecole (environments, observation functions etc.) is designed with mixed-integer linear optimization in mind. Things might work with non-linear optimization as well, although I'm not sure what the expected behaviour should be. Regarding strong branching scores, I am not sure if those even have a standard definition for nonlinear optimization. Maybe this is something you should ask to the SCIP developers directly.
>
> Best,
> Maxime

**Comment by AntoinePrv** [comment]
Hi @cloud-waiting-for-wind,

Unfortunately, we do not know as much about SCIP as we'd like to. 😅 It's a complex software that we do not develop ourselves, and very often in Ecole, we do not go behind the hood of what is going on in SCIP.

In theory, strong branching could be applied to any branch and bound method (with any way to compute a lower bound). However, I don't know if that's the case with the one in SCIP.
@gasse you've actually implemented that `VanillaFullStrongBranching` that is used here. Do you not know if it was restricted to MILP?

**Reply by gasse** [reply]
I'm not sure. To compute SB scores we call existing methods SCIPgetVarStrongbranchInt() and SCIPgetVarStrongbranchFrac(), so the question is whether those methods are restricted to MILPs or not. See our implementation here:
https://www.scipopt.org/doc/html/branch__vanillafullstrong_8c_source.php
